ex1/dates_times.py

```diff
@@ -32,13 +32,39 @@
 
 # Define a function named last_of_month that takes an argument dt of type date
 # and returns a date object representing the last day of the month dt was in.
-
+import calendar, datetime
+def last_of_month(dt):
+	if dt.month == 1 or dt.month == 3 or dt.month == 5 or dt.month == 7 or dt.month == 8 or dt.month == 10 or dt.month == 12:
+		return datetime.date(dt.year, dt.month, 31)
+	elif dt.month == 2 and calendar.isleap(dt.year) == True:
+		return datetime.date(dt.year, dt.month, 29)
+	elif dt. month == 2 and calendar.isleap(dt.year) == False:
+		return datetime.date(dt.year, dt.month, 28)
+	elif dt.month == 4 or dt.month == 6 or dt.month == 9 or dt.month == 11:
+		return datetime.date(dt.year, dt.month, 30)
 # Define a function named feed_the_gremlin which takes a time object as an
 # argument. It should return False between midnight and 6:30AM and True
 # otherwise.
+def feed_the_gremlin(tm):
+	if tm.hour >= 6 and tm.minute >= 30:
+		return True
+	else:
+		return False
 
 # Define a function named how_long that takes two datetime objects dt and ref
 # where ref is the reference datetime, calculates the difference between them and
 # returns the difference as a string formatted like:
 # "01 days, 01 minutes, 01 seconds until 2000-12-31 15:59:59"
 # If ref is before dt then use 'since' instead of 'until'
+from datetime import timedelta
+def how_long(dt, ref):
+	if (dt < ref):
+		delta = ref - dt
+		minu = delta.seconds//60
+		second = delta.seconds - minu*60
+		return "{} days, {} minutes, {} seconds until {}" .format(delta.days, minu, second, ref)
```

Comment:
In general you should use `datetime.datetime.strftime` for this kind of formatting.

In this case it is lucky that
```python
>> "{}".format(datetime(2007,1,1,15,35,12))
'2007-01-01 15:35:12'
```
works automatically.